pandas/io/common.py

```diff
@@ -126,9 +127,17 @@ def _validate_header_arg(header) -> None:
         )
 
 
-def _stringify_path(
-    filepath_or_buffer: FilePathOrBuffer[AnyStr]
-) -> FilePathOrBuffer[AnyStr]:
+@overload
```

Comment:
https://stackoverflow.com/questions/37553551/3-5-type-hinting-and-method-overloading

is there a reason we *need* overload here?

Reply:
we could be less strict and use casts or ignores instead where necessary

Reply:
this seems to be hiding a bug then, these signatures *should* all be the same no? why do we *need* different signatures?

Reply:
the signatures are the same. the types are different.

buf passed through unchanged (as typevar)

Path converted to string

so overload is so that if a Path is given a string is returned.

Reply:
I see so this is essentially doing a Union of 2 TypeVars then. ok, can you give a couple of comments before the use of overload then to explain what is happening. my hesitation is that the next person to look at this will not understand w/o some hints.

Reply:
This is a great application of overload - thanks @simonjayhawkins

Reply:
To see if I understand correctly, overloading allows us to properly type e.g.:

```
def foo(x, y):
   if isinstance(x, Foo) and isinstnace(y, Bar):
        [do stuff]
   elif isinstance(x, Bar) and isinstnace(y, Foo):
        [do stuff]
   else:
        raise TypeError(x, y)
```

where if we just typed it as `Union[Foo, Bar]` for both `x` and `y` mypy would infer that Bar, Bar was OK when it isn't.  And overload allows us to specify the exact combinations that are OK.

Is that close?

Reply:
yep.

Reply:
```python
from typing import overload


class Foo:
    pass


class Bar:
    pass


@overload
def foo(x: Foo, y: Bar) -> str:
    ...


@overload
def foo(x: Bar, y: Foo) -> int:
    ...


def foo(x, y):
    if isinstance(x, Foo) and isinstance(y, Bar):
        return "foo"
    elif isinstance(x, Bar) and isinstance(y, Foo):
        return 42
    else:
        raise TypeError(x, y)


foo(Foo(), Bar())
foo(Bar(), Foo())
foo(Foo(), Foo())
foo(Bar(), Bar())
```
```
$ mypy test.py
test.py:33: error: No overload variant of "foo" matches argument types "Foo", "Foo"
test.py:33: note: Possible overload variants:
test.py:33: note:     def foo(x: Foo, y: Bar) -> str
test.py:33: note:     def foo(x: Bar, y: Foo) -> int
test.py:34: error: No overload variant of "foo" matches argument types "Bar", "Bar"
test.py:34: note: Possible overload variants:
test.py:34: note:     def foo(x: Foo, y: Bar) -> str
test.py:34: note:     def foo(x: Bar, y: Foo) -> int
```

Reply:
> I see so this is essentially doing a Union of 2 TypeVars then. ok, can you give a couple of comments before the use of overload then to explain what is happening. my hesitation is that the next person to look at this will not understand w/o some hints.

@jreback I imagine we could be using overload quite a bit. I'll add comments if you think necessary, but it may end up adding quite a bit of duplication with the docstrings over time.

Reply:
maybe add a small section in the contributing docs then of why / how we are using it?

Reply:
Contributing Guide for Type Hints #27050 is not yet merged. added comments for now using comments from https://mypy.readthedocs.io/en/latest/more_types.html#function-overloading as boilerplate.